src/components/Api.js

```diff
@@ -0,0 +1,143 @@
+export default class API {
+  constructor({ baseUrl, headers }) {
+    // constructor body
+    this._baseUrl = baseUrl;
+    this._headers = headers;
+  }
+
+  _handleResponse(res) {
+    if (res.ok) {
+      // checks the server response
+      return res.json();
+    } else {
+      // if server returns an error, reject the promise
+      return Promise.reject(`Error ${res.status}`);
+    }
+  }
+
+  // Card routes
+  // GET /cards – Get all cards
+  getInitialCards() {
+    return fetch(`${this._baseUrl}/cards`, {
+      headers: this._headers,
+    }).then(this._handleResponse);
```

Comment:
**COULD BE IMPROVED**

You can make a special method (function ) for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._handleResponse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

Reply:
Hi Gennady, I resubmitted the updated code.  For the special method (_request) how would it need to be written with template literals not having to duplicate in every request?

Reply:
> Hi Gennady, I resubmitted the updated code. For the special method (_request) how would it need to be written with template literals not having to duplicate in every request?

Hi)

**COULD BE IMPROVED**

You can make a universal method `_request` not to duplicate the `ok` checking, `headers` and `baseUrl`


```js
 _request(endpoint, options = {}) {
    const finalOptions = {
      // add the default headers
      headers: this._headers,
      // spread the other options if provided
      ...options,
    };
      // add the endpoint to the base url
    const url = `${this._baseUrl}${endpoint}`;
    return fetch(url, finalOptions).then(this._checkResponse);
  }

  // usage

  getUserInfo() {
    // pass only the endpoint here
    return this._request("/users/me");
  }

  getInitialCards() {
    // pass only the endpoint here
    return this._request("/cards");
  }

  // and so on



```

Reply:
For this special method, why is options ={} a parameter? Also, how does this work for addCard({name, link}) or updateAvatar(url) - which have different method, headers, and body property?

Reply:
> For this special method, why is options ={} a parameter?

If you don't pass `options`, they'll be an empty object by default. It's a feature of js

Reply:
is _request(endpoint, options) the same as _request(endpoint, options = {}) ?